sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSession.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.ServiceBus.Amqp;
+using Azure.Messaging.ServiceBus.Core;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Amqp.Framing;
+
+namespace Azure.Messaging.ServiceBus
+{
+    /// <summary>
+    ///
+    /// </summary>
+    public class ServiceBusSession
+    {
+        private readonly TransportConsumer _consumer;
+        private readonly ServiceBusRetryPolicy _retryPolicy;
+        private readonly bool _isSessionful;
+
+        internal ServiceBusSession(
+            TransportConsumer consumer,
+            ServiceBusRetryPolicy retryPolicy,
+            bool isSessionful)
+        {
+            _consumer = consumer;
+            _retryPolicy = retryPolicy;
+            _isSessionful = isSessionful;
+        }
+
+        /// <summary>
+        /// TODO implement
+        /// </summary>
+        /// <returns></returns>
+        public virtual async Task<byte[]> GetStateAsync(
+            CancellationToken cancellationToken = default)
+        {
+            return await Task.FromResult(new byte[4]).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// TODO implement
+        /// </summary>
+        /// <param name="sessionState"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public virtual async Task SetStateAsync(byte[] sessionState,
+            CancellationToken cancellationToken = default)
+        {
+            await Task.Delay(1).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// TODO implement
+        /// </summary>
+        /// <returns></returns>
+        public virtual async Task RenewSessionLockAsync(CancellationToken cancellationToken = default)
+        {
+            await Task.Delay(1).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public virtual async Task<string> GetSessionId(CancellationToken cancellationToken = default)
+        {
+            if (!_isSessionful)
+            {
+                return null;
+            }
+            ReceivingAmqpLink openedLink = null;
+            await _retryPolicy.RunOperation(
+                async () =>
+                openedLink = await _consumer.ReceiveLink.GetOrCreateAsync(_retryPolicy.Options.TryTimeout).ConfigureAwait(false),
```

Comment:
This confuses me.   The point of the transport consumer is to abstract the concept of AMQP links away from callers, especially those outside of the AMQP layer.   In its current form , this is leaking the abstraction.   This makes it much harder to inject a transport consumer for behavioral overrides to unit test and to potentially add other transports (like a recording framework) in the future that may not be link-based.

I could be missing something, but shouldn't we be having the details of getting the session on the transport consumer or another transport client?    I'd also advise against exposing the link directly, to avoid leaking that abstraction and force the consumer to be extended instead.

Reply:
The implementation can be moved down to the consumer layer.